src/common/Formatter.h

```diff
@@ -54,6 +55,122 @@ namespace ceph {
       }
     };
 
+    /// Helper to check if a type is a map for our purpose
+    /// (based on fmt code)
+    template <typename T> class is_map {
+      template <typename U> static auto check(U*) -> typename U::mapped_type;
+      template <typename> static void check(...);
+    public:
+      static constexpr const bool value =
+        !std::is_void<decltype(check<T>(nullptr))>::value;
+    };
+
+    /**
+     * with_array_section()
+     * Opens an array section and calls 'fn' on each element in the container.
+     * Two overloads are provided:
+     * 1. for maps, where the function takes a key and a value, and
+     * 2. for other types of containers, where the function takes just an
+     *    element.
+     */
+
+    // for maps
+    template <
+	typename M,     //!< a map<K, V>
+	typename FN,    //!< a callable to be applied to each element
+	typename K = std::remove_cvref_t<M>::key_type,
+	typename V = std::remove_cvref_t<M>::mapped_type>
+      requires(
+	  is_map<M>::value && (
+          std::is_invocable_v<FN, Formatter&, const K&, const V&, std::string_view> ||
+          std::is_invocable_v<FN, Formatter&, const K&, const V&>))
+    void with_array_section(std::string_view txt, const M& m, FN&& fn) {
+      Formatter::ArraySection as(*this, txt);
+      for (const auto& [k, v] : m) {
+        if constexpr (std::is_invocable_v<FN, Formatter&, const K&, const V&, std::string_view>) {
+	  std::invoke(std::forward<FN>(fn), *this, k, v, txt);
+        } else {
+          std::invoke(std::forward<FN>(fn), *this, k, v);
```

Comment:
probably we can start using C++23? so we can have the luxury of using ranges library in Ceph?

Reply:
AFAIK: we have not made the official move to C++23.
(and the min clang is 16 still, according to the cmakelists from a couple of days ago)